components/application-broker/internal/storage/populator/instances.go

```diff
@@ -69,11 +93,132 @@ func (p *Instances) Do(ctx context.Context) error {
 	for _, si := range serviceInstances {
 		if si.Spec.ServiceClassRef != nil {
 			if _, ex := abClassNames[si.Spec.ServiceClassRef.Name]; ex {
-				if err := p.inserter.Insert(p.converter.MapServiceInstance(si)); err != nil {
-					return errors.Wrap(err, "while inserting service instance")
+				p.log.Infof("process ServiceInstance (%s)", si.Spec.ExternalID)
+				if err := p.restoreInstanceData(si); err != nil {
+					return errors.Wrap(err, "while saving service instance data")
 				}
 			}
 		}
 	}
 	return nil
 }
+
+func (p *Instances) restoreInstanceData(si *v1beta1.ServiceInstance) error {
+	var addInstance bool
+
+	switch p.specifyRestoreMode(si) {
+	case readyMode:
+		p.log.Info("ServiceInstance is in ready state")
+		_, err := p.addOperation(si, internal.OperationTypeCreate, internal.OperationStateSucceeded, "")
+		if err != nil {
+			return errors.Wrap(err, "while add operation for ready instance")
+		}
+		addInstance = true
+	case provisionPollingLastOperationMode:
+		p.log.Info("ServiceInstance is in failed provision state with PollingLastOperation error")
+		var params map[string]interface{}
+		if si.Spec.Parameters != nil {
+			err := json.Unmarshal(si.Spec.Parameters.Raw, &params)
+			if err != nil {
+				return errors.Wrap(err, "while unmarshaling instance parameters")
+			}
+		}
+		p.log.Info("restore provisioning process")
+		err := p.broker.ProvisionProcess(broker.RestoreProvisionRequest{
+			Parameters:           params,
+			InstanceID:           internal.InstanceID(si.Spec.ExternalID),
+			OperationID:          internal.OperationID(*si.Status.LastOperation),
+			Namespace:            internal.Namespace(si.Namespace),
+			ApplicationServiceID: internal.ApplicationServiceID(si.Spec.ServiceClassRef.Name),
+		})
+		if err != nil {
+			return errors.Wrap(err, "while triggering provisioning process")
+		}
+
+		_, err = p.addOperation(si, internal.OperationTypeCreate, internal.OperationStateInProgress, internal.OperationID(*si.Status.LastOperation))
+		if err != nil {
+			return errors.Wrap(err, "while add operation for failed instance (PollingLastOperation error)")
+		}
+		addInstance = true
+	case deprovisionPollingLastOperationMode, deprovisionCallFailedMode:
+		p.log.Info("ServiceInstance is in failed deprovision state")
+		_, err := p.addOperation(si, internal.OperationTypeCreate, internal.OperationStateSucceeded, "")
+		if err != nil {
+			return errors.Wrap(err, "while add operation for failed instance (DeprovisionCallFailed)")
+		}
+		opID, err := p.addOperation(si, internal.OperationTypeRemove, internal.OperationStateInProgress, "")
+		if err != nil {
+			return errors.Wrap(err, "while add operation for failed instance (DeprovisionCallFailed)")
+		}
+		p.broker.DeprovisionProcess(broker.DeprovisionProcessRequest{
+			Instance:             p.converter.MapServiceInstance(si),
+			OperationID:          opID,
+			ApplicationServiceID: internal.ApplicationServiceID(si.Spec.ServiceClassRef.Name),
```

Comment:
**[BLOCKER] needs to be investigated before merging it to master**

did you test that with api-packages?  IMO may not work properly

Here you have a helper method to select a proper id:
https://github.com/kyma-project/kyma/blob/f2c3b3498f91c22250ddbc7a6a4449b679a40263/components/application-broker/internal/broker/broker.go#L201-L227

I was using it in other OSB API calls. The problem is that in the new approach:
- `appID` = `classID`
- `serviceID` = `planID`

in the old approach:
- `serviceID` = `classID`
- `hardcoded-id`  = `planID`


Please check that and if changes will be required please also add a test case to highlight that contract and to prevent breaking it in the future.

Reply:
fixed, please take a look